### PR TITLE
feat: add slippage warning to MenuButton

### PR DESCRIPTION
**Pull request by grabbou**
## Description

Adds Slippage warning to the Menu button as per Figma. Also, fixes the following:
- correct top/bottom margins and horizontal paddings of the SwapHeader
- correct top margin of the MenuFlyout (aligned with the SwapInput)

Other less significant changes:
- Extracts slippage validation to a separate utility `validateUserSlippageTolerance` for reusability
- Removes a bunch of extraneous and unnecessary styling properties (made 3 years ago, likely by moving them from Figma)

Ticket: https://linear.app/uniswap/issue/WEB-2041/add-slippage-badge-to-swap-component

| Before       | After (Desktop) | 
| ------------ |---------------- | 
| Screenshot 2023-05-12 at 22 18 54 | Screenshot 2023-05-12 at 22 16 27 |
| | Screenshot 2023-05-12 at 22 16 20 |
| | Screenshot 2023-05-12 at 22 16 12 |
| Screenshot 2023-05-12 at 22 19 01 | Screenshot 2023-05-12 at 22 29 29

### CODE REVIEW

File: src/components/Settings/index.tsx [hunk 1]
```diff
@@ -90,37 +53,29 @@ export default function SettingsTab({ autoSlippage }: { autoSlippage: Percent })
...
-      >
-        <StyledMenuIcon data-testid="swap-settings-button" />
-      </StyledMenuButton>
-      {open && (
+    <Menu ref={node}>
+      <MenuButton disabled={!isSupportedChain} isActive={isOpen} onClick={toggleMenu} />
+      {isOpen && (
         <MenuFlyout>
-          <AutoColumn gap="16px" style={{ padding: '1rem' }}>
-            {isSupportedChainId(chainId) && <RouterPreferenceSettings />}
```
Comment (grabbou) [inline src/components/Settings/index.tsx:112]: note: this check was unnecessary, because `MenuButton` was always disabled for unsupported chains, essentially making this menu never appear (`isOpen = false`).

File: src/components/Settings/index.tsx [hunk 2]
```diff
@@ -1,80 +1,43 @@
...
   box-shadow: 0px 0px 1px rgba(0, 0, 0, 0.01), 0px 4px 8px rgba(0, 0, 0, 0.04), 0px 16px 24px rgba(0, 0, 0, 0.04),
     0px 24px 32px rgba(0, 0, 0, 0.01);
   border-radius: 12px;
-  display: flex;
-  flex-direction: column;
-  font-size: 1rem;
   position: absolute;
-  top: 2rem;
-  right: 0rem;
+  top: 100%;
```
Comment (tinaszheng) [inline src/components/Settings/index.tsx:30]: :catjam:

File: src/state/user/hooks.tsx [hunk 3]
```diff
@@ -149,10 +149,16 @@ export function useUserSlippageTolerance(): [
     [dispatch]
   )
 
-  return useMemo(
```
Comment (tinaszheng) [inline src/state/user/hooks.tsx:152]: lmao

File: src/components/Settings/MenuButton/index.tsx [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+  width: 20px;
+  > * {
+    stroke: ${({ theme }) => theme.textSecondary};
+  }
+`
+
+const Button = styled.button<{ isActive: boolean }>`
+  border: none;
+  background-color: transparent;
+  margin: 0;
```
Comment (tinaszheng) [inline src/components/Settings/MenuButton/index.tsx:21]: isnt this default? 🤔

File: src/components/Settings/MenuButton/index.tsx [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+        <Icon />
+      </IconContainer>
+    )
+  }
+
+  const isInvalidSlippage = validateUserSlippageTolerance(userSlippageTolerance) !== SlippageValidationResult.Valid
+
+  return (
+    <IconContainerWithSlippage gap="sm" displayWarning={isInvalidSlippage}>
+      <ThemedText.Caption>{userSlippageTolerance.toFixed(2)}% slippage</ThemedText.Caption>
```
Comment (tinaszheng) [inline src/components/Settings/MenuButton/index.tsx:62]: `slippage` should be translated

File: src/components/Settings/MenuButton/index.test.tsx [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+    expect(renderButton()).toMatchSnapshot()
+  })
+  it('should render an icon with a custom slippage value', () => {
+    mocked(useUserSlippageTolerance).mockReturnValue([new Percent(5, 10_000), noop])
+    expect(renderButton()).toMatchSnapshot()
+  })
+  it('should render an icon with a custom slippage and a warning when value is out of bounds', () => {
+    mocked(useUserSlippageTolerance).mockReturnValue([new Percent(1, 10_000), noop])
+    expect(renderButton()).toMatchSnapshot()
+  })
```
Comment (zzmp) [inline src/components/Settings/MenuButton/index.test.tsx:29]: These tests should check for what they're expecting, not just check against a snapshot.
Snapshots are most useful to check that nothings changed, but here you're explicitly expecting that there is `a custom slippage value...`.

File: src/utils/validateUserSlippageTolerance.ts [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+
+export default function validateUserSlippageTolerance(userSlippageTolerance: Percent) {
+  switch (true) {
+    case userSlippageTolerance.lessThan(MINIMUM_RECOMMENDED_SLIPPAGE):
+      return SlippageValidationResult.TooLow
+    case userSlippageTolerance.greaterThan(MAXIMUM_RECOMMENDED_SLIPPAGE):
+      return SlippageValidationResult.TooHigh
+    default:
+      return SlippageValidationResult.Valid
+  }
```
Comment (zzmp) [inline src/utils/validateUserSlippageTolerance.ts:20]: Why use a `switch` instead of an if/else statement? I think this'd be more readable and idiomatic with if/else.

File: src/utils/validateUserSlippageTolerance.ts [hunk 8]
```diff
@@ -0,0 +1,21 @@
+import { Percent } from '@uniswap/sdk-core'
+
+export enum SlippageValidationResult {
+  TooLow = 'TOO_LOW',
+  TooHigh = 'TOO_HIGH',
+  Valid = 'VALID',
```
Comment (zzmp) [inline src/utils/validateUserSlippageTolerance.ts:6]: nit: I thought we named enums in all caps, eg `TOO_LOW`.

File: src/utils/validateUserSlippageTolerance.ts [hunk 9]
```diff
@@ -0,0 +1,21 @@
+import { Percent } from '@uniswap/sdk-core'
+
+export enum SlippageValidationResult {
```
Comment (zzmp) [inline src/utils/validateUserSlippageTolerance.ts:3]: These are enums, so you don't need to give them values - you can just let the compiler default-assign them.

File: src/utils/validateUserSlippageTolerance.test.ts [hunk 10]
```diff
@@ -0,0 +1,15 @@
+import { Percent } from '@uniswap/sdk-core'
+
+import validateUserSlippageTolerance, { SlippageValidationResult } from './validateUserSlippageTolerance'
+
+describe('validateUserSlippageTolerance', () => {
```
Comment (zzmp) [inline src/utils/validateUserSlippageTolerance.test.ts:5]: Can you add the edge cases where it is _equal to_ the minimum/maximum? These seem the most likely to break on a potential future refactor.

File: src/state/user/hooks.tsx [hunk 11]
```diff
@@ -149,10 +149,16 @@ export function useUserSlippageTolerance(): [
...
-    () => [userSlippageTolerance, setUserSlippageTolerance],
-    [setUserSlippageTolerance, userSlippageTolerance]
-  )
+  return [userSlippageTolerance, setUserSlippageTolerance]
+}
+
+/**
+ * Same as above but replaces the auto with a default value
```
Comment (zzmp) [inline src/state/user/hooks.tsx:156]:
```suggestion
 * Returns user slippage tolerance, replacing the auto with a default value.
```

File: src/components/Settings/MenuButton/index.tsx [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+const IconContainerWithSlippage = styled(IconContainer)<{ displayWarning?: boolean }>`
+  div {
+    color: ${({ theme, displayWarning }) => (displayWarning ? theme.accentWarning : theme.textSecondary)};
+  }
+
+  background-color: ${({ theme, displayWarning }) =>
+    displayWarning ? theme.accentWarningSoft : theme.backgroundModule};
+`
+
+const ButtonIcon = () => {
```
Comment (zzmp) [inline src/components/Settings/MenuButton/index.tsx:47]: nit: `ButtonContent`, as this covers more than just the icon.

File: src/components/Settings/MenuButton/index.tsx (hunk 5 above)
Comment (zzmp) [reply src/components/Settings/MenuButton/index.tsx:62]: The entire caption should be translated, not just the word, as some locales may change the order of number/word.

**Review by zzmp — CHANGES_REQUESTED** [review]
great cleanup around the feature! looks good, just some requests for changes around testing/style.

File: src/utils/validateUserSlippageTolerance.ts (hunk 8 above)
Comment (grabbou) [reply src/utils/validateUserSlippageTolerance.ts:6]: I am not aware of that styling convention, searching the codebase we have a mix of both. We may want to standardise in the future.

File: src/utils/validateUserSlippageTolerance.test.ts (hunk 10 above)
Comment (grabbou) [reply src/utils/validateUserSlippageTolerance.test.ts:5]: Done

File: src/utils/validateUserSlippageTolerance.ts (hunk 7 above)
Comment (grabbou) [reply src/utils/validateUserSlippageTolerance.ts:20]: Personal preference, I think both are equally readable. However, having searched the codebase, I think if/else is more established, so I will adjust my preference to match the codebase.

File: src/components/Settings/MenuButton/index.tsx (hunk 4 above)
Comment (grabbou) [reply src/components/Settings/MenuButton/index.tsx:21]: Unfortunately, it isn't. CSS reset isn't going anywhere lol

**Review by tinaszheng — APPROVED** [review]
!!

File: src/state/user/hooks.tsx [hunk 13]
```diff
@@ -149,10 +149,16 @@ export function useUserSlippageTolerance(): [
...
 
-  return useMemo(
-    () => [userSlippageTolerance, setUserSlippageTolerance],
-    [setUserSlippageTolerance, userSlippageTolerance]
-  )
+  return [userSlippageTolerance, setUserSlippageTolerance]
+}
+
+/**
+ *Returns user slippage tolerance, replacing the auto with a default value
```
Comment (zzmp) [inline src/state/user/hooks.tsx:156]:
```suggestion
 * Returns user slippage tolerance, replacing the auto with a default value
```

**Review by zzmp — APPROVED** [review]
